### DISCUSSION TITLE
Get Object.Label of Managed Metadata

### DISCUSSION THREAD

**Discussion by LegendaryB** [pnp/pnpcore#595] [Q&A] [answered]
Hey,

I feel like a guest who comes back every week. So sorry about it xD

When I use this approach (item is `IListItem`):
```csharp
if (item.Properties.Values.TryGetValue(propertyName, out var value))
    document.Properties.Add(propertyName, value);
```
I get this value for a managed metadata property: `1;#Foo|b7044609-0f6a-4105-9721-76feec6416d4`
But I'm interested in FieldValue.Label. Any way to get that?

**Comment by jansenbe** [comment, marked as answer]
@LegendaryB , when reading the list holding the items I assume you've also loaded the list fields as mentioned here https://pnp.github.io/pnpcore/using-the-sdk/listitems-intro.html#reading-list-items? If so then a managed metadata field should be represented by an `IFieldTaxonomyValue` or an `IFieldValueCollection` of `IFieldTaxonomyValue` for multi taxonomy fields. Which should give you the taxonomy label and id (https://pnp.github.io/pnpcore/api/PnP.Core.Model.SharePoint.IFieldTaxonomyValue.html).

Now you can also simply split the string yourselves (label = Foo, Id = b704...)...

**Reply by LegendaryB** [reply]
@jansenbe 
Yeah I load the list like this. But I don't understand how I can now get the Label of the single `IListItem` instance? In the properties of it I have the value but as a concatenated string

**Reply by LegendaryB** [reply]
Or do you mean just splitting that string on char "#" and "|" 

Because this will result in the Label right?

**Reply by jansenbe** [reply]
@LegendaryB : Or you use the typed field value class to get the label:
![image](https://user-images.githubusercontent.com/7451219/138669718-f98b1502-aa65-4d9c-bbb3-04ae30f9b43d.png)

**Reply by LegendaryB** [reply]
@jansenbe Wanted to try your approach but only get this exception:
```
fail: SPODocumentDistribution.App[0]
      HttpResponseCode: 400
      Code: Microsoft.SharePoint.SPException
      Message: The field or property 'Values' does not exist.
      ClientRequestId:
```
I query for values like this:
```
list.Items
     .QueryProperties(p => p.Values)
     .AsAsyncEnumerable();
```

**Reply by jansenbe** [reply]
You don't need to query "Values", it's always there as that's where the list item properties are stored. They're stored in this dictionary as we support a generic IListItem which needs to work for any type of list/library

**Reply by LegendaryB** [reply]
@jansenbe  At my side the collection is always empty when I look at `IListItem.Values` so I thought I would need to query it before using.

**Reply by jansenbe** [reply]
@LegendaryB : hmm aren't we/me mixing up the list item properties with the list item property bag...they're totally different concepts...can you clarify what you want to use?

**Reply by LegendaryB** [reply]
@jansenbe 
My current task is to find all documents -> download them and also write a output CSV which should contain some basic information about the sharepoint ListItem. For example Name, Version, Unique Id. This stuff is pretty easy to get. But I also need to retrieve field values as a string. So like in my example from above. Our field is called "company_test_mms" and returns "1;#Foo|b7044609-0f6a-4105-9721-76feec6416d4". But I'm only after the Foo in this case.

**Reply by LegendaryB** [reply]
So you could say I want the string representation of it. On boolean it would be simply true or false for example

**Reply by jansenbe** [reply]
@LegendaryB : So you're getting the property bag of a document, not the actual metadata set on the list holding the document. These are two completely different things, typically document metadata is set as custom fields on the list holding the document and not as properties in the document property bag. I would suggest to cross check your source libraries and understand what you really need. Using https://github.com/pnp/pnpcore/discussions/586#discussioncomment-1496214 is what you need for list properties, you just need to specify the ones you need in the CAML query of drop the viewfields part to get all properties

**Reply by LegendaryB** [reply]
@jansenbe 
Yes correct I use the `IListItem.Properties` property at the moment to get for example `company_test_mms`. Maybe I'm mixing something up but I thought I need to get it from there because not every `IListItem` has a value set for column `company_test_mms`. Because it's not flagged as required

**Reply by jansenbe** [reply]
@LegendaryB : What you see is the metadata that lives inside the (office) document (check the properties of your file). When a doc is added to a library that data is copied to the property bag. Might be that all docs had this property (they were copied from a SharePoint on-prem location?), but that's not a given as other docs in the library might not. If there was a metadata column in the document library with the same name as the property in the document that column will be automatically populated.

![image](https://user-images.githubusercontent.com/7451219/138697939-963b7c51-1683-43c9-ae08-f8f6cece7085.png)

Dialog shown are the properties listed via Word

**Reply by LegendaryB** [reply]
@jansenbe 
The documents are coming just from my local machine. I generated them via a script.
It's okay when not all documents have values for the properties. I just search for a way to query a string value for the property when it is there. 

For example I want to get the Label for a Taxonomy property, true/false for boolean, Text normal string, Multi Line text also as a string (without breaks) and so on. But I guess I need some logic on my side to handle all of these scenarios, right? There is not something like: property.Value.ToString() => done? :D

**Reply by LegendaryB** [reply]
@jansenbe 
My only problem still left is that `IListItem.Values` is empty. Do you have an idea why when it is something which should be filled by default? 

Edit: Maybe it has something todo with the fact that I use `AsAsyncEnumerable`? When I use a snippet with the extension `AsRequested` the values are filled.

**Reply by jansenbe** [reply]
@LegendaryB : can you paste the relevant code snippet?

**Reply by LegendaryB** [reply]
@jansenbe 
Sorry about the formatting, straight out of Visual Studio^^:
```csharp
var itemsEnumerable = list.Items
                .QueryProperties(
                    p => p.UniqueId,
                    p => p.Properties,
                    p => p.FileSystemObjectType,
                    p => p.File)
                .AsAsyncEnumerable();

            await foreach (var item in itemsEnumerable)
            {
                _logger.LogInformation($"Starting to process list item: {item.UniqueId}");

                _logger.LogInformation($"Values count: {item.Values.Count}");

                // processing...

                _logger.LogInformation($"Finished processing of list item: {item.UniqueId}");
            }
```

**Reply by jansenbe** [reply]
@LegendaryB : can you also show how you load the list?

**Reply by LegendaryB** [reply]
@jansenbe 
```csharp
 using var context = await _pnpContextFactory.CreateAsync(site);

            var listsEnumerable = context.Web.Lists
                .Where(l => l.TemplateType == ListTemplateType.DocumentLibrary &&
                       l.Hidden == false &&
                       l.IsSystemList == false &&
                       l.IsSiteAssetsLibrary == false &&
                       l.IsCatalog == false)
                .QueryProperties(
                    p => p.Title, 
                    p => p.Items,
                    p => p.Fields.QueryProperties(
                        p => p.InternalName, 
                        p => p.FieldTypeKind, 
                        p => p.TypeAsString, 
                        p => p.Title))
                .AsAsyncEnumerable();
```

**Reply by jansenbe** [reply]
@LegendaryB : when you load the list you already load all the items in the list...and later on you're loading the items again. So either you use the items you loaded when loading the list, to do so just use the .AsRequested() method and do a for each or LINQ query on the items. Alternatively you don't load the items with the list but load them separately per library. If you do that then use the `All` property. See below example:

```csharp
var list = await context.Web.Lists.GetByTitleAsync("LargeVersion");

var itemsEnumerable = list.Items
.QueryProperties(
    p => p.All,
    p => p.Properties,
    p => p.FileSystemObjectType,
    p => p.File)
.AsAsyncEnumerable();

await foreach (var item in itemsEnumerable)
{
    // Get property from property bag
    var myCustomPropViaProperties = item.Properties.GetString("MyCustomProp", null);
    // Get property from list metadata
    var myCustomPropViaListMetadata = item.Values["MyCustomProp"]?.ToString();
}
```

**Reply by LegendaryB** [reply]
Thanks @jansenbe,

I'm now using this approach:
```csharp
var lists = context.Web.Lists
.Where(
    l => l.TemplateType == ListTemplateType.DocumentLibrary &&
    l.Hidden == false &&
    l.IsSystemList == false &&
    l.IsSiteAssetsLibrary == false &&
    l.IsCatalog == false);

foreach (var list in lists) 
{
    var itemsEnumerable = list.Items
         .QueryProperties(
             p => p.All,
             p => p.UniqueId,
             p => p.Properties,
             p => p.FileSystemObjectType,
             p => p.File)
         .AsAsyncEnumerable();

    await foreach (var item in itemsEnumerable)
    {
         _logger.LogInformation($"Starting to process list item: {item.UniqueId}");

         // further processing...
    }
}
```
Does this look like you described or have you any other tips? Thanks! 

One question left: How can I now retrieve a complex object in the Values for example `FieldUserValue`? Its there but Value Property is only partially loaded.

**Reply by jansenbe** [reply]
These complex properties are loaded based upon the available information returned from the request + the availability of the list field information (which was not in your code). So add a `QueryProperties` to your list loading like shown below.

```csharp
var lists = context.Web.Lists
                        .QueryProperties(
                            p => p.Title,
                            // Add other list fields you need
                            //..
                            p => p.Fields.QueryProperties(p => p.InternalName,
                                                                   p => p.FieldTypeKind,
                                                                   p => p.TypeAsString,
                                                                   p => p.Title))
                        .Where(
                            l => l.TemplateType == ListTemplateType.DocumentLibrary &&
                            l.Hidden == false &&
                            l.IsSystemList == false &&
                            l.IsSiteAssetsLibrary == false &&
                            l.IsCatalog == false);

await foreach (var list in lists.AsAsyncEnumerable())
{
    var itemsEnumerable = list.Items
         .QueryProperties(
             p => p.All,
             p => p.UniqueId,
             p => p.Properties,
             p => p.FileSystemObjectType,
             p => p.File)
         .AsAsyncEnumerable();

    await foreach (var item in itemsEnumerable)
    {
        

        // further processing...
    }
}
```

Since SharePoint, in case of a user, only returns the id of the user for the current site that's the only thing that get's loaded up. You can't automatically get more user data via adding additional QueryProperties, you'll need to load that separately if you need that. 

@LegendaryB : just noticed that the Field property is null here, checking that...

![image](https://user-images.githubusercontent.com/7451219/138827532-1d5c4f37-b218-4a52-b050-bb433ffb38a0.png)

**Reply by LegendaryB** [reply]
Thank you for checking. Looks the same on my side 👍

**Reply by jansenbe** [reply]
@LegendaryB : The fact that the field property is null is something we should fix, but it will not impact your scenario. 

There's 2 ways to get SharePoint ListItem data, using the "regular" get approach you're using or using a CAML query (see https://github.com/pnp/pnpcore/discussions/586#discussioncomment-1496214). Both have pro's and con's:

Option | Pro | Con
----|----|----
Regular approach | Can load `Properties` in one go | Most system list item properties are not returned
CAML query models | All properties returned, more detail for some properties (like user's) | Loading `Properties` will require a separate call per list item, so really expensive

As you only need custom added metadata and you need the file's `Properties` the regular approach seems to be the best for you, so the code shown in the previous reply.

**Reply by LegendaryB** [reply]
@jansenbe 
Yeah I also guess that this would be the best solution for my scenario. But then I need to do manual splitting right? Because a user object is then something like: 1#Name 
But should be okay I guess. Needs then a little bit more testing but yeah :D

**Reply by jansenbe** [reply]
Maybe also let's mark this one as answered :-)
